tools/coqdoc/output.ml

```diff
@@ -107,6 +107,50 @@ let set_module m sub = current_module := (m,sub);
 let item_level = ref 0
 let in_doc = ref false
 
+let (%) f g x = f (g x)
+
+let string_flatmapi f =
+  let b = Buffer.create 16 in
+  fun s ->
+    Buffer.clear b;
+    String.iteri (fun i c -> Buffer.add_string b (f i c)) s;
+    Buffer.contents b
+
+let string_flatmap f = string_flatmapi (fun _ -> f)
+
+let html_escape = string_flatmap (function
+  | '<' -> "&lt;"
+  | '>' -> "&gt;"
+  | '&' -> "&amp;"
+  | '\"' -> "&quot;"
+  | c -> String.make 1 c
+  )
+
+let uri_escape = string_flatmap (function
+  | '>' | '<' | '#' | '&' as c -> sprintf "%%%X" (int_of_char c)
+  | ' ' -> "_"
+  | c -> String.make 1 c
+  )
+
+(* A 'fragment' is the part in the uri path that comes after the '#'. *)
+
+let allowed_fragment_char = function
+  | 'a'..'z' | 'A'..'Z' | '0'..'9' | '.' | '_' | '<' | '>' | '&'
+    | '\'' | '"' | '-' | ':' -> true
```

Comment:
This leads to the following (notice the `%3E` part).
```latex
\coqexternalref{http://coq.inria.fr/stdlib/Coq.Init.Logic}{::type\_scope:x\_'-\%3E'\_x}{\coqdocnotation{\ensuremath{\rightarrow}}}
```

Either you allow only pure fragment characters (and use `Digest.string` for nonconforming fragments) or you encode every forbidden character as `%XX` (and never use `Digest.string`). But please do not use a mix of both. (I have a small preference for the former, i.e., removing `<`, `>` from the list above.)

By the way, the list seems incomplete. I think `/`, `?`, `:`, `@`, `~`, `!`, `$`, `(`, `)`, `*`, `+`, `,`, `;`, `=` are also allowed. Where they omitted on purpose? (Backward compatibility?)